shark/src/main/java/shark/internal/PathFinder.kt

```diff
@@ -462,6 +473,51 @@ internal class PathFinder(
     }
   }
 
+  private fun HeapInstance.readAllNonNullFieldsOfReferenceType(): MutableList<HeapField> {
+    val hprofGraph = graph // DO NOT REMOVE! Getter performs typecast which is expensive!
+    var fieldReader: FieldValuesReader? = null
+    val result = mutableListOf<HeapField>()
+    var skipBytesCount = 0
+    for (heapClass in instanceClass.classHierarchy.toList()) {
+      for (fieldRecord in heapClass.readRecord().fields) {
+        if (fieldRecord.type != PrimitiveType.REFERENCE_HPROF_TYPE) {
+          // Skip all fields that are not references. Track how many bytes to skip
+          skipBytesCount += hprofGraph.getRecordSize(fieldRecord)
+        } else {
+          // Initialize field reader if it's not yet initialized. Replaces `lazy` without synchronization
+          if (fieldReader == null) {
+            fieldReader = hprofGraph.createFieldValuesReader(readRecord())
+          }
+
+          // Skip the accumulated bytes offset
+          fieldReader.skipBytes(skipBytesCount)
+          skipBytesCount = 0
+
+          val fieldValue = ReferenceHolder(fieldReader.readId())
+          if (!fieldValue.isNull) {
+            val fieldName = hprofGraph.fieldName(heapClass.objectId, fieldRecord.nameStringId)
```

Comment:
Replace this with a new API: `HeapClass.fieldName(fieldRecord: FieldRecord)` or `HeapClass.fieldName(fieldIndex: Int)`

That way we don't have to make hprofGraph.fieldName public.